src/storage/ExtensionBasedMapper.ts

```diff
@@ -137,7 +137,7 @@ export class ExtensionBasedMapper implements FileIdentifierMapper {
     let relative = filePath.slice(this.rootFilepath.length);
     if (isContainer) {
       return {
-        identifier: { path: encodeURI(this.baseRequestURI + relative) },
+        identifier: { path: ensureTrailingSlash(encodeURI(this.baseRequestURI + relative)) },
```

Comment:
I'd expect the base URI to already be encoded, and the relative part to require an `encodeURIComponent`.

Reply:
`encodeURI` is easier because `encodeURIComponent` would require us to split all parts again and encode every part separately. `encodeURI` also encodes all non-alphanumeric characters, even the ones that are valid in an URL. E.g., `encodeURI(';,/?:@&=+$') === ';,/?:@&=+$'`, but `encodeURIComponent(';,/?:@&=+$') === '%3B%2C%2F%3F%3A%40%26%3D%2B%24'`.

So it comes down to what we want to have as "canonical" URL for a resource internally to make sure we stay consistent. Currently the incoming URL gets fully decoded in the the `TargetExtractor` and encoded again with `encodeURI`, thereby making sure the URL is still a valid URL. That also means that afterwards we can just use `decodeURI` and `encodeURI` to change back and forth.

The solution with `encodeURIComponent` could also work, that would require us to first fully decode any incoming URL and then completely encode it again (instead of just using `encodeURI`), it's just important to be consistent to make sure we always use the same identifier for the same reason.

With the curent solution, an incoming URL like `http://test.com/$%24/` would be converted to `http://test.com/$$/`  internally. If we used `encodeURI` it would be `http://test.com/%24%24/`. I think it's nicer if we use the first one, but the most important thing remains that we never have it as `http://test.com/$%24/` internally.

Reply:
Good arguments.

However, we should be careful with `encodeURI` in general, to the point I usually _never_ use it (unless there are very valid reasons; the ones above might be those). It is kind-of intended as a "fixer" method: that is, pass it an improperly escaped URL, and it will fix it. The issue is that the improperly escaped URL should have never existed in the first place: it is not needed if all parts of the URL were properly encoded with `encodeURIComponent` (which is not a fixer method, but rather a full escaping/unescaping mechanism for _parts_ of URIs).

In this case, I am specifically concerned about `encodeURI` not escaping `# + & = @`, which could appear in file-based backends, and in general also `? :`, which could appear in other backends.

Reply:
So concretely, I would propose to make

> split all parts again and encode every part separately

a utility method `uriEncodePathComponents`.

Reply:
Yea fair point, it could produce weird URLs, I'll change it like that. Besides potential esthetics I can't really think of any immediate disadvantages.